nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java

```
@@ -0,0 +1,322 @@
+  }
+
+  @Override
+  public boolean dropTable(TableIdentifier identifier, boolean purge) {
+    reference.checkMutable();
+
+    IcebergTable existingTable = table(identifier);
+    if (existingTable == null) {
+      return false;
+    }
+
+    // We try to drop the table. Simple retry after ref update.
+    int count = 0;
+    while (count < 5) {
+      count++;
+      try {
+        dropTableInner(identifier);
+        break;
+      } catch (NessieConflictException e) {
+        // pass for retry
+      } catch (NessieNotFoundException e) {
+        logger.error("Cannot drop table: ref is no longer valid.", e);
+        return false;
+      }
+    }
+    if (count >= 5) {
+      logger.error("Cannot drop table: failed after retry (update hash and retry)");
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  public void renameTable(TableIdentifier from, TableIdentifier toOriginal) {
+    reference.checkMutable();
+
+    TableIdentifier to = NessieUtil.removeCatalogName(toOriginal, name());
+
+    IcebergTable existingFromTable = table(from);
+    if (existingFromTable == null) {
+      throw new NoSuchTableException("table %s doesn't exists", from.name());
+    }
+    IcebergTable existingToTable = table(to);
+    if (existingToTable != null) {
+      throw new AlreadyExistsException("table %s already exists", to.name());
+    }
+
+    Operations contents = ImmutableOperations.builder()
+        .addOperations(ImmutablePut.builder().key(NessieUtil.toKey(to)).contents(existingFromTable).build(),
+            ImmutableDelete.builder().key(NessieUtil.toKey(from)).build())
+        .build();
+
+    try {
+      client.getTreeApi().commitMultipleOperations(reference.getAsBranch().getName(), reference.getHash(),
+          "iceberg rename table", contents);
+      // TODO: fix this so we don't depend on it in tests.
+      refresh();
+    } catch (NessieConflictException e) {
+      throw new CommitFailedException(e, "failed");
```

I'd prefer an error message with more context, like `Failed to rename X to Y`. There's no guarantee that the error message from Nessie has that information. It probably has information about the state of the ref, rather than what was being attempted.

---

yup. I believe I was lazy here b/c that exception is always swallowed by the re-try logic above. Not such a good excuse :-) fixed now